File: python/30-39/problem34.py
from math import factorial


def digit_factorials(n):
    """Returns a list of all digit_factorials up to ``n``.

    Args:
        n: Max number to check for digit factorials.

    Returns:
        List of digit factorials.
    """
    return [x for x in range(3, n + 1) if is_digit_factorial(x)]


def is_digit_factorial(n):
    """Checks if a numer is a digit factorial.

    Args:
        n: Number to check if it is a factorial.

    Returns:
        True if ``n`` is a digit factorial. False if not.
    """
    digit_factorial = sum([factorial(int(x)) for x in str(n)])
    return digit_factorial == n



if __name__ == '__main__':
    print(sum(digit_factorials(100000)))

File: js/30-39/problem34.js
const prototypes = require('../prototypes');

// Returns the factorial of a number without using recursion.
var factorial = function(n) {
    if (n == 0) {
        return 1;
    }
    let product = n;
    for (let i = n - 1; i > 0; i--) {
        product = product * i;
    }
    return product;
};

// Returns an array of the factorials of the digits in 'n'.
var factorialArray = function(n) {
    let answer = [];
    for (let i of (Number.intToArray(n))) {
        answer.push(factorial(i))
    }
    return answer;
};

/* Returns an array of numbers that equal the sum of the factorials of their
digits up to 'max'. */
var digitFactorials = function(max) {
    let answer = [];
    for (let i = 3; i <= max; i++) {
        if (i === factorialArray(i).sum()) {
            answer.push(factorialArray(i).sum())
        }
    }
    return answer;
};

console.log(digitFactorials(100000).sum());
